index.bs

```diff
@@ -424,58 +433,150 @@ When this method is invoked, the user agent MUST execute the following algorithm
+    : {{hashAlg}}
+    :: UA-chosen
+
+        Issue(w3c/webauthn#362): We need *some* constraints on the possible hash
+        algorithms, or else sites will fail on unusual UAs.
+    : {{tokenBinding}}
+    :: The [=Token Binding ID=] associated with |callerOrigin| (if any)
+
+        Issue(w3c/webauthn#360): Make sure this association was set up properly.
+    : {{ClientData/extensions}}
+    :: |clientExtensions|
+
+1. Let |clientDataJSON| be the [=UTF-8 encoding=] of the result of calling the
+    initial value of {{JSON/stringify|JSON.stringify}} on |clientData|.
+
+    Issue: Some extensions contain ArrayBuffers, which don't stringify well.
+    What's the intent here?
+1. Let |clientDataHash| be the hash of |clientDataJSON| using
+    <code>|clientData|.{{hashAlg}}</code>.
+
+1. Let |issuedRequests| and |currentlyAvailableAuthenticators| be new [=ordered
+    sets=].
+
+1. For each |authenticator| currently available on this platform, if
+    <code>{{options}}.{{ScopedCredentialOptions/attachment}}</code> is not
+    [=present=] or its value matches |authenticator|'s attachment modality,
+    [=set/append=] |authenticator| to |currentlyAvailableAuthenticators|.
+
+1. [=set/For each=] |authenticator| in |currentlyAvailableAuthenticators|:
+    1. Let |excludeList| be a new [=list=].
+    1. [=list/For each=] credential |C| in <code>{{options}}.{{ScopedCredentialOptions/excludeList}}</code>:
+        1. If |C| has an empty {{transports}} list, [=list/append=] |C| to
+            |excludeList| and [=continue=].
+        1. If |authenticator| is connected over a transport not mentioned in
+            <code>|C|.{{transports}}</code>, the client MAY [=continue=].
+
+            Issue: I'm not sure this captures the intent of the original wording.
+        1. [=list/Append=] |C| to |excludeList|.
+    1. [=In parallel=], invoke the <a>authenticatorMakeCredential</a> operation
+        on |authenticator| with |rpId|, |clientDataHash|,
+        {{accountInformation}}, |normalizedParameters|, |excludeList| and
+        |clientExtensions| as parameters.
+    1. [=set/Append=] |authenticator| to |issuedRequests|.
+
+1. Let |promise| be [=a new promise=]. Return |promise| and start a timer for |adjustedTimeout| milliseconds.
+    Then execute the following steps [=in parallel=]. If any fatal error is encountered in this algorithm other than the
+    ones enumerated below, cancel the timer, [=reject=] |promise| with a DOMException whose name is "{{UnknownError}}", and terminate
+    this algorithm.
 
-1. Use {{attestationChallenge}}, |callerOrigin| and |rpId|, along with the token binding key associated with |callerOrigin| (if
-    any), to create a {{ClientData}} structure representing this request. Choose a hash algorithm for {{ClientData/hashAlg}} and
-    compute the {{ScopedCredentialInfo/clientDataJSON}} and its <a>clientDataHash</a>.
+    Issue: What kinds of fatal errors are you worried about? I suggest we just
+    remove that sentence.
```

Comment:
OK. The initial intention was to guard against things like low-level communication breakdowns between client and authenticator, but maybe that becomes part of "any authenticator returns an error" now.